pkg/openstackclient/client.go

```diff
@@ -0,0 +1,307 @@
+		return nil, fmt.Errorf("failed to create server: %w", err)
+	}
+	return server, nil
+}
+
+// Image methods
+func (c *openstackClient) GetImage(ctx context.Context, imageID string) (*images.Image, error) {
+	return images.Get(c.imageClient, imageID).Extract()
+}
+
+// Ports
+func (c *openstackClient) ListPorts(ctx context.Context) ([]ports.Port, error) {
+	allPages, err := ports.List(c.networkClient, nil).AllPages()
+	if err != nil {
+		return nil, err
+	}
+
+	return ports.ExtractPorts(allPages)
+}
+
+// Network specific
+func (c *openstackClient) GetNetworkByName(ctx context.Context, networkName string) (*networks.Network, error) {
+	listOpts := networks.ListOpts{
+		Name: networkName,
+	}
+
+	allPages, err := networks.List(c.networkClient, listOpts).AllPages()
+	if err != nil {
+		return nil, err
+	}
+
+	networkList, err := networks.ExtractNetworks(allPages)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(networkList) == 0 {
+		return nil, fmt.Errorf("network with name '%s' not found", networkName)
+	}
+
+	// Return the first match directly
+	return &networkList[0], nil
+}
+
+// Get the security group names for a specific server
+func (c *openstackClient) GetServerSecurityGroupNames(ctx context.Context, serverID string) ([]string, error) {
+	serverSecGroups, err := secgroups.ListByServer(c.computeClient, serverID).AllPages()
+	if err != nil {
+		return nil, err
+	}
+
+	secGroupList, err := secgroups.ExtractSecurityGroups(serverSecGroups)
+	if err != nil {
+		return nil, err
+	}
+
+	var secGroupNames []string
+	for _, secGroup := range secGroupList {
+		secGroupNames = append(secGroupNames, secGroup.Name)
+	}
+
+	return secGroupNames, nil
+}
+
+// Create a client wrapper object for interacting with OpenStack.
+func NewClientFromSecret(secret *corev1.Secret) (Client, error) {
+	cloudsYaml, ok := secret.Data[constants.OpenStackCredentialsName]
+	if !ok {
+		return nil, fmt.Errorf("secret does not contain %q data", constants.OpenStackCredentialsName)
+	}
+
+	// Use upstream structs instead of custom ones
+	var clouds clientconfig.Clouds
+	if err := yaml.Unmarshal(cloudsYaml, &clouds); err != nil {
+		return nil, fmt.Errorf("failed to parse clouds.yaml: %w", err)
+	}
+
+	// Get the first cloud (or specific cloud name)
+	var cloud *clientconfig.Cloud
+	for _, c := range clouds.Clouds {
+		cloud = &c
+		break
+	}
+
+	if cloud == nil {
+		return nil, fmt.Errorf("no cloud configuration found in clouds.yaml")
+	}
+
+	authInfo := &clientconfig.AuthInfo{
+		AuthURL:           cloud.AuthInfo.AuthURL,
+		Username:          cloud.AuthInfo.Username,
+		Password:          cloud.AuthInfo.Password,
+		ProjectID:         cloud.AuthInfo.ProjectID,
+		ProjectName:       cloud.AuthInfo.ProjectName,
+		UserDomainName:    cloud.AuthInfo.UserDomainName,
+		ProjectDomainName: cloud.AuthInfo.ProjectDomainName,
+	}
+
+	opts := &clientconfig.ClientOpts{
+		AuthInfo:     authInfo,
+		RegionName:   cloud.RegionName,
+		EndpointType: cloud.Interface,
+	}
```

Comment:
I'm curious why this can't be
```suggestion
	opts := &clientconfig.ClientOpts{
		AuthInfo:     cloud.AuthInfo,
		RegionName:   cloud.RegionName,
		EndpointType: cloud.Interface,
	}
```

Does it have something to do with needing to scrub out the other fields of `cloud.AuthInfo`?